File: intel-hex-to-bin.scala/src/main/scala/demo/Hex.scala
package demo

/**
  * Converters to hexadecimal strings.
  */
object Hex {

  def to8BitHexString(value: Short): String = {
    require((0 to 0xFF).contains(value))
    "0x%02X".format(value)
  }

  def to16BitHexString(value: Int): String = {
    require((0 to 0xFFFF).contains(value))
    "0x%04X".format(value)
  }

  def to32BitHexString(value: Int): String = {
    require(value >= 0)
    "0x%04X".format(value)
  }

  def to8BitHexString(values: Array[Short]): String = {
    val s = new StringBuilder()
    s.append('[')
    for (i <- values.indices) {
      val value = values(i)
      s.append(Hex.to8BitHexString(value))
      if (i < values.length - 1) {
        s.append(",")
      }
    }
    s.append(']').toString()
  }

}

File: intel-hex-to-bin.java/src/main/java/demo/Hex.java
package demo;

import static com.google.common.base.Preconditions.checkArgument;

/**
 * Converters from numbers to hexadecimal strings.
 */
public interface Hex {

    static String to8BitHexString(short value) {
        checkArgument((value >= 0) && (value <= 0xFF));
        return String.format("0x%02X", value);
    }

    static String to16BitHexString(int value) {
        checkArgument((value >= 0) && (value <= 0xFFFF));
        return String.format("0x%04X", value);
    }

    static String to32BitHexString(int value) {
        checkArgument(value >= 0);
        return String.format("0x%04X", value);
    }

    static String to8BitHexString(short[] values) {
        StringBuilder s = new StringBuilder();
        s.append('[');
        for (int i = 0; i < values.length; i++) {
            s.append(to8BitHexString(values[i]));
            if (i < values.length - 1) {
                s.append(",");
            }
        }
        return s.append(']').toString();
    }

}
